### PR TITLE
[Kiali] changes for the next version

### PR DESCRIPTION
**Pull request by jmazzitelli**

### CODE REVIEW

File: install/kubernetes/helm/subcharts/kiali/templates/deployment.yaml [hunk 1]
```diff
@@ -48,11 +48,13 @@ spec:
             secretKeyRef:
               name: {{ .Values.dashboard.secretName }}
               key: {{ .Values.dashboard.usernameKey }}
+              optional: true
```
Comment (sdake) [inline install/kubernetes/helm/subcharts/kiali/templates/deployment.yaml]: I RTD at https://kubernetes.io/docs/concepts/configuration/secret/ - however, it is a little slim on details of the symantics of the optional secretKeyRef.  Will this result in a mounting if the secrets if they are dynamically created?

The symantics I believe we agreed upon were to use created secrets if they were present, if not present to start up but not provide any access to the system.  Once the secreets are registered, the secrets would then be mounted into the deployment and be usable.

Does this PR do that?

**Comment by sdake** [comment]
Can you please rebase - we fixed a failure in MCP testing about 1 day ago that may make the gate work better.

Cheers
-steve

File: install/kubernetes/helm/subcharts/kiali/templates/deployment.yaml (hunk 1 above)
Comment (jmazzitelli) [reply install/kubernetes/helm/subcharts/kiali/templates/deployment.yaml]: For background on why this "optional=true" was added to the secret-backed env var definitions, see https://github.com/istio/istio/issues/11244 and [KIALI-2303](https://issues.jboss.org/browse/KIALI-2303)

What happens is: IF the secret does NOT exist, the pod will still start, however the environment variables based on the secret will NOT be defined. Kiali will detect this condition and will not provide any access to the system - instead, the user will get [an error message in the login screen](https://issues.jboss.org/secure/attachment/12446051/Screenshot%20from%202019-01-25%2015-44-47.png) indicating the problem. Once the secret is created, the pod is required to be restarted (e.g. `kubectl delete pod --selector=app=kiali` will cause the pod to be scaled up back to 1 and access will now be granted assuming the secret exists and the user provides the valid credentials at login time). I do not know of a way to get kubernetes to auto-restart a pod if it detects a secret that was missing has been created during the lifetime of a pod.

**Comment by jmazzitelli** [comment]
> I am curious if we have met all of our goals with the secrets work for Kiali. Could you clarify?

At minimum, I can say no user will gain access to Kiali if the secret is missing. So from a security standpoint, I believe we are good. Are we expecting pods to restart automatically when creation of a previously missing secret is detected by kubernetes? It doesn't do that (manual restart is required) but I do not know of a way for kubernetes to do that - for env. vars that are backed by a secret, a started pod would need to be re-started to pick up new secrets.

**Comment by jmazzitelli** [comment]
> Can you please rebase - we fixed a failure in MCP testing about 1 day ago that may make the gate work better.

I just checked and this is already based on the tip of master. I just restarted the tests -- let's see what happens.

File: install/kubernetes/helm/subcharts/kiali/templates/deployment.yaml (hunk 1 above)
Comment (sdake) [reply install/kubernetes/helm/subcharts/kiali/templates/deployment.yaml]: @jmazzitelli sounds like a good interrmediate solution.  I also don't know of any way to do this with Kubuernetes either using manifests.  We have used a solution like the following to solve this problem though with go code:

Watch for secret registration with the correct name in the correct namespace - if it is present, then use it.  Check out:
https://github.com/istio/istio/tree/master/pkg/kube/secretcontroller

For an example.

Cheers
-steve

**Comment by sdake** [comment]
@jmazzitelli check my review comment related to the secretcontroller written by @sbezverk for istio. 
 thanks for all the hard work!

/lgtm
/approve

**Comment by jmazzitelli** [comment]
> check my review comment related to the secretcontroller written by @sbezverk for istio.

I created an enhancement request for Kiali here: https://issues.jboss.org/browse/KIALI-2397

**Pull request merged**